testcases/kernel/syscalls/recvmmsg/recvmmsg01.c

```diff
@@ -0,0 +1,2250 @@
+static void con_make_timeout_tests(struct con *conp);
+static void con_init_socketpair(struct con *conp);
+static void con_init_seqpacket(struct con *conp);
+static void con_init_dgram(struct con *conp);
+static void con_deinit(struct con *conp);
+static void con_sendmmsg_recvmmsg(struct con *conp, size_t n,
+				  int sflags, struct mmsghdr *smm,
+				  int rflags, struct mmsghdr *rmm);
+static void con_do_multi_send_recv(struct con *conp);
+static void con_add_send_recv_calls_vec(struct con *conp,
+					struct sendmsg_call *smcp,
+					struct recvmsg_call *rmcp,
+					bool sender_sleeps);
+static void con_add_send_recv_calls(struct con *conp,
+				    struct sendmsg_call *smcp,
+				    struct recvmsg_call *rmcp,
+				    bool sender_sleeps);
+
+static int call_go(struct call *callp, struct con *conp);
+
+static int call_receive_iovec_boundary_checks(struct call *callp,
+					      struct con *conp, int tn);
+static int call_receive_iovec_boundary_checks_peeking(struct call *callp,
+						      struct con *conp, int tn);
+static int call_receive_file_descriptor(struct call *callp,
+					struct con *conp, int tn);
+static int call_message_too_long_to_fit_discards_bytes(struct call *callp,
+						       struct con *conp,
+						       int tn);
+static int call_message_too_long_to_fit_doesnt_discard_bytes(struct call *callp,
+							     struct con *conp,
+							     int tn);
+static int call_receive_waits_for_message(struct call *callp,
+					  struct con *conp, int tn);
+static int call_receiver_doesnt_wait_for_messages(struct call *callp,
+						  struct con *conp, int tn);
+static int call_receive_returns_what_is_available(struct call *callp,
+						  struct con *conp, int tn);
+static int call_empty_datagram_received_as_empty_datagram(struct call *callp,
+							  struct con *conp,
+							  int tn);
+
+static int con_receive_iovec_boundary_checks(struct con *conp,
+		int tn, size_t niov);
+static int con_receive_iovec_boundary_checks_peeking(struct con *conp, int tn,
+						     size_t niov);
+static int con_receive_file_descriptor(struct con *conp, int tn,
+				       int cloexec_flag,
+				       bool some_data,
+				       ssize_t controllen_delta);
+static int con_message_too_long_to_fit_discards_bytes(struct con *conp, int tn);
+static int con_message_too_long_to_fit_doesnt_discard_bytes(struct con *conp,
+							    int tn);
+static int con_receive_waits_for_message(struct con *conp, int tn);
+static int con_receiver_doesnt_wait_for_messages(struct con *conp, int tn);
+static int con_receive_returns_what_is_available(struct con *conp, int tn);
+static int con_empty_datagram_received_as_empty_datagram(struct con *conp,
+							 int tn);
+
+/*
+ * test_assert() checks something that should not fail in the test,
+ * if it fails the test code is broken, not the system call being tested.
+ */
+
+#define	test_assert(expr) 						\
+	do {								\
+		if (! (expr)) {						\
+			tst_res(TBROK, "%s", #expr);			\
+			exit(1);					\
```

Comment:
This is still wrong, the test should do tst_brk(TBROK, ...) that will exit the test instead of tst_res() and exit(1).